File: java/src/main/java/de/sven_torben/serialization_benchmark/serializer/AbstractSerializer.java
package de.sven_torben.serialization_benchmark.serializer;

import java.io.InputStream;
import java.io.OutputStream;

public abstract class AbstractSerializer<T> implements ISerializer<T> {

	private final Class<T> type;
	
	public AbstractSerializer(final Class<T> type)
	{
		this.type = type;
	}
	
	@Override
	public abstract void serialize(OutputStream stream, T data) throws Exception;
	
	@Override
	public abstract T deserialize(InputStream stream) throws Exception;

	@Override
	public String getName() {
		return this.getClass().getSimpleName();
	}

	@Override
	public final Class<T> getType() {
		return type;
	}
	
}

File: net/Serializer/AbstractSerializer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace De.Sven_Torben.Serialization_Benchmark.Serializer
{
    abstract class AbstractSerializer<T> : ISerializer<T>
    {

        protected AbstractSerializer(T testSample)
        {
            TestSample = testSample;
        }

        public virtual String Name
        {
            get { return this.GetType().Name; }
        }

        public T TestSample
        {
            get;
            protected set;
        }


        public abstract void Serialize(System.IO.Stream stream, T data);

        public abstract T Deserialize(System.IO.Stream stream);
    }
}
